### ISSUE TITLE
Persistent session filter not working as expected from 1.19.x onwards

### ISSUE THREAD

**Issue by cybrhwk** [istio/istio#49000]
### Is this the right place to submit this?

- [X] This is not a security vulnerability or a crashing bug
- [X] This is not a question about how to use Istio

### Bug Description

I have implemented the persistent session filter as my application is stateful and requires sticky sessions to function correctly and the following istio setup works perfectly between 1.17.2 & 1.18.7. From 1.19.x the configuration no longer works, including the latest 1.20.x version.

For the working versions the session remains "sticky" for hours, however it lasts just over a minute when using 1.19.x onwards, at which time I'm logged out as the last request appears to have been sent to a different pod.

I have ran istioctl analyze and it hasn't shown anything wrong with the configuration.

![istio-flow-diagram drawio](https://github.com/istio/istio/assets/16922631/f1b62163-e01b-41f1-96b9-a4637432759f)

### Version

```Text
client version: 1.19.6
control plane version: 1.19.6
data plane version: 1.19.6 (62 proxies)

Client Version: v1.26.0
Kustomize Version: v4.5.7
Server Version: v1.25.16-eks-8cb36c9
```

**Comment by howardjohn** [comment]
cc @costinm

**Comment by costinm** [comment]
Thanks John. The generated config is same, so we may have a bug on envoy side.

**Comment by costinm** [comment]
Can you include sample responses ( with headers ) ? Is the cookie not set, or does it change ? 

The cookie is a base64 encoded IP:port of an endpoint - it would be worth checking the IP is present in the EDS ( istioctl pc endopoints).

**Comment by ramaraochavali** [comment]
Can you share the config dump of those two versions?

**Comment by cybrhwk** [comment]
The output of running 'istioctl pc endpoints' is below:

ENDPOINT                   STATUS        OUTLIER CHECK     CLUSTER
100.64.17x.167:7001    HEALTHY     OK                             outbound|7001||my-app.my-namespace.svc.cluster.local
100.64.17x.167:9000    HEALTHY     OK                             outbound|9000||my-app.my-namespace.svc.cluster.local
100.64.17x.171:7001    HEALTHY     OK                             outbound|7001||my-app.my-namespace.svc.cluster.local
100.64.17x.171:9000    HEALTHY     OK                             outbound|9000||my-app.my-namespace.svc.cluster.local
100.64.6x.217:7001      HEALTHY     OK                             outbound|7001||my-app.my-namespace.svc.cluster.local
100.64.6x.217:9000      HEALTHY     OK                             outbound|9000||my-app.my-namespace.svc.cluster.local
100.64.6x.221:7001      HEALTHY     OK                             outbound|7001||my-app.my-namespace.svc.cluster.local
100.64.6x.221:9000      HEALTHY     OK                             outbound|9000||my-app.my-namespace.svc.cluster.local

I've attached the headers from initial login and then the subsequent logout re-login having been moved to another pod. The value of the stickysession cookie changes when this happens

[istio-initial-login-headers.txt](https://github.com/istio/istio/files/14064161/istio-initial-login-headers.txt)
[istio-initial-logout-after-short-time-login-headers.txt](https://github.com/istio/istio/files/14064160/istio-initial-logout-after-short-time-login-headers.txt)

**Comment by cybrhwk** [comment]
@ramaraochavali What do you mean by config dump? What versions do you require the config from? 

The configuration I'm running to create the environment remains the same between a working sticky session and one that isn't working, the only thing I am changing is the istio version

**Comment by costinm** [comment]
@ramaraochavali - we have another customer reporting this, and I checked the config dumps and it is identical.

**Comment by costinm** [comment]
I wonder if the fact there are 2 cookies (servlet and envoy) is creating any problems. Thanks for the headers - they also confirm that envoy is adding a cookie - and it appears to have the right content. 

I see: JSESSIONID=DDFD49E7EB75DD13AB648884F7573C32; stickysession="ChIxMDAuNjQuNjkuMjIxOjcwMDEQniM="; 
and JSESSIONID=6C67A079C3D9E3780D6F0FD7BD38C1DC; stickysession="ChIxMDAuNjQuNjkuMjE3OjcwMDEQ2CQ="

so the traffic moved from .221 to .217 - and both appear healthy. 

The session filter is clearly doing something - adding the cookie, the question is why does it not stick.

**Comment by costinm** [comment]
Thanks to Kuat for reminding me - could you check Istiod or envoy logs for any XDS pushes ? We want to rule out any listener (or other) config update - i.e. to check if the stickysession moves to a different value in the absence of any config changes.

**Comment by cybrhwk** [comment]
@costinm Do I need to configure anything to enable the logging for the XDS pushes?

**Comment by costinm** [comment]
No, pretty sure istiod and envoy log something on push.

**Comment by ramaraochavali** [comment]
Could be related https://github.com/envoyproxy/envoy/pull/26761 - May be we should raise an issue with Envoy?

**Comment by cybrhwk** [comment]
> I wonder if the fact there are 2 cookies (servlet and envoy) is creating any problems. Thanks for the headers - they also confirm that envoy is adding a cookie - and it appears to have the right content.
> 
> I see: JSESSIONID=DDFD49E7EB75DD13AB648884F7573C32; stickysession="ChIxMDAuNjQuNjkuMjIxOjcwMDEQniM="; and JSESSIONID=6C67A079C3D9E3780D6F0FD7BD38C1DC; stickysession="ChIxMDAuNjQuNjkuMjE3OjcwMDEQ2CQ="
> 
> so the traffic moved from .221 to .217 - and both appear healthy.
> 
> The session filter is clearly doing something - adding the cookie, the question is why does it not stick.

@costinm This JSESSIONID cookie is also present in 1.17.2 which behaves correctly. Do you still require the logs to look for XDS pushs?

@ramaraochavali This certainly looks like it could be the cause, I guess the EnvoyFilter mechanism doesn't update the cookie ttl held in the Envoy proxies, so Envoy goes with the default 120s ttl? Can the configuration of the ttl for the stateful session filter in istio be done soon? That would remove the need for the EnvoyFilter

**Comment by cybrhwk** [comment]
> Thanks to Kuat for reminding me - could you check Istiod or envoy logs for any XDS pushes ? We want to rule out any listener (or other) config update - i.e. to check if the stickysession moves to a different value in the absence of any config changes.

I can see XDS updates happening in the proxy logs, but they look like the routine updates. There doesn't seem to be any correlation between these update messages happening and the sticky session ending either immediately after it happens or within a few seconds of it occurring.

**Comment by cybrhwk** [comment]
> Could be related [envoyproxy/envoy#26761](https://github.com/envoyproxy/envoy/pull/26761) - May be we should raise an issue with Envoy?

@ramaraochavali This certainly looks like it could be the cause, I guess the EnvoyFilter mechanism doesn't update the cookie ttl held in the Envoy proxies, so Envoy goes with the initial default 120s ttl? 

Can the configuration of the ttl for the stateful session filter cookie in istio be done soon? 

Do I raise another issue for that feature request? That would remove the need for the EnvoyFilter

**Comment by costinm** [comment]
We can certainly set a larger TTL. 

I am not entirely sure how to expose 'persistent cookie' vs 'session cookie', but we can certainly make changes to the Istiod to avoid the use of the envoy filter. 

To get an idea of what you need - what value would you use, and would it be persistent cookie or session ?

**Comment by costinm** [comment]
I can confirm the cookies are protobuf - when I first looked I didn't pay attention to the 'garbage' in the output,
 
 echo ChIxMDAuNjQuNjkuMjE3OjcwMDEQ2CQ= | base64 -d | od -t ax1
0000000  nl dc2   1   0   0   .   6   4   .   6   9   .   2   1   7   :
         0a  12  31  30  30  2e  36  34  2e  36  39  2e  32  31  37  3a
0000020   7   0   0   1 dle   X   $
         37  30  30  31  10  d8  24

**Comment by costinm** [comment]
@kyessenov - I think best short term option is to change envoy so that if ttl=0, we preserve the documented behavior and don't add an expiration in the cookie (and don't expire cookies without expiration ). 

We may refine the API in the future - but first step should be to fix the code.

**Comment by costinm** [comment]
And on Istiod side - how about adding a new label for the 'ttl', for consistency - until we implement the upstream K8S API for persistent sessions ?   I see the upstream API has AbsoluteTimeoutSeconds and IdleTimeoutSeconds - but still doesn't make any difference between session cookies and persistent cookies.

I am not sure if the default should be 0 ( session cookie ) or 172800s - the chance that an endpoint will stay the same for that long is extremely low, and defaulting to session cookies (ttl=0) may be a better fit ( and may not need the GDPR stuff in browsers ? )

**Comment by costinm** [comment]
https://gdpr.eu/cookies/ for reference...

**Comment by costinm** [comment]
Another likely bug (based on the headers and config): the cookie does not have "Expires" attribute, despite ttl being set to 120s. The other bug we are investigating has ttl=0, so explicit request for a session cookie. 

I opened a bug on gateway API too - https://github.com/kubernetes-sigs/gateway-api/issues/2747

**Comment by cybrhwk** [comment]
> And on Istiod side - how about adding a new label for the 'ttl', for consistency - until we implement the upstream K8S API for persistent sessions ? I see the upstream API has AbsoluteTimeoutSeconds and IdleTimeoutSeconds - but still doesn't make any difference between session cookies and persistent cookies.
> 
> I am not sure if the default should be 0 ( session cookie ) or 172800s - the chance that an endpoint will stay the same for that long is extremely low, and defaulting to session cookies (ttl=0) may be a better fit ( and may not need the GDPR stuff in browsers ? )

Yeah another label makes sense since one is already used to enable the stateful session filter already.

Would I be right in thinking, if the longer ttl is set at the creation point instead of being modified later by the EnvoyFilter, the issue with Envoy encoding the expiration inside the cookie wouldn't be an issue or would the Envoy issue also have to be resolved for it all to hook together and function as it does on 1.17.2 and 1.18.7?

**Comment by costinm** [comment]
I'm not sure. IMHO the simplest option is to default to 'session' cookies (ttl=0), and perhaps to not support "Expires" (browser stored cookies ) at all for the use cases of stickiness to an endpoint. 

Sessions for SSO or user tracking obviously need to have support for long expiration and real persistence (permanent cookies) - but that's a different feature from 'send to same endpoint'. On the other hand - the upstream API in K8S Gateway seems to mention   the Expires header and pretty complicated expiration, and supporting it will be trickier.

**Comment by cybrhwk** [comment]
Do you have an idea on when a build of istio will be available to test the various fixes linked to this issue? I would like to upgrade to the latest version of istio as soon as possible, since at the moment I am stuck on 1.18.X, which is out of support

**Comment by costinm** [comment]
Envoy PR needs to merge first, than after we update envoy version in Istio
my PR.

I am hoping to backport this at least to current version (if rel managers
agree).

**Comment by cybrhwk** [comment]
Apart from the change from having a cookie with an expiry to using a session cookie for providing the persistent session functionality, will there be any other differences to how the mechanism functions in 1.17.x and 1.18.x I should be aware of?

**Comment by costinm** [comment]
No, same label and behavior.

We will eventually add support for the new K8S API - but no reason to
remove the label and it's not a priority right now, the k8s API is
still evolving.

One advice: don't rely on the fact the cookie encodes the IP:port, envoy
changed the format once and at some point we may
encrypt the cookie value. So treat it as an opaque.

**Comment by cybrhwk** [comment]
@costinm I just wondered if there was any news on once envoy publish their new version if that new version would be backported to the current 1.20.x version of istio?

**Comment by costinm** [comment]
Kuat made the envoy change - but release managers need to decide. It is not a beta/stable feature yet ( in envoy or Istio ), but is pretty small and safe.

Envoy backport and picking the fixed version goes first.

**Comment by costinm** [comment]
It's a question for Kuat and release managers.

**Comment by cybrhwk** [comment]
@costinm Would you be able to bring it to Kuat's attention for me please?

**Comment by costinm** [comment]
I talked with Kuat - my understanding is that the fix was not backported, it is not a security issue. I think it is always best to use the last version of Istio - there are a lot of other fixes and improvements, and skip-version upgrade generally works.

**Comment by cybrhwk** [comment]
@costinm I completely agree and I am happy to move to whichever version the fixes are in since I would like to use latest and greatest, I guess my question then is, will that mean it'll be in 1.22.x of istio at the earliest and not in a minor release of 1.21.x?

**Comment by costinm** [comment]
As of now - yes. I don't necessarily like this - but that's the process.

If you have good arguments ( like impacting production, version that works out of support window, etc) it is reasonable to ask the release managers - but someone will need to backport the other envoy fix first. 

Both the EnvoyFilter approach and the annotation are not 'v1'/stable - since the underlying feature in Envoy is not stable ( as is obvious by this bug and their own docs/description), so we can't treat it as a critical fix. 

I hope that once the K8S Gateway API is implemented by other envoy-based projects we'll have more pressure on envoy to mark this as stable and support it.

**Comment by cybrhwk** [comment]
The version that works being out of the support window is my primary concern, as it will become a compliance issue eventually, so the sooner I can get off the version I'm on the better. I would imagine other people that rely on persistent sessions will be in the same boat as me.

Is there a roadmap that indicates when 1.22.x will be out? If it's fairly soon then that would probably just about work, if its months away then it wouldn't
